src/coreclr/jit/codegenarm64.cpp

```diff
@@ -4412,57 +4412,166 @@ void CodeGen::genCodeForCompare(GenTreeOp* tree)
+        emit->emitIns_R_I_FLAGS_COND(INS_ccmp, cmpSize, srcReg1, (int)intConst->IconValue(), cflags, cond);
     }
     else
     {
-        assert(!varTypeIsFloating(op2Type));
-        // We don't support swapping op1 and op2 to generate cmp reg, imm.
-        assert(!op1->isContainedIntOrIImmed());
-        // This should not be generating into a register.
-        assert(targetReg == REG_NA);
+        regNumber srcReg2 = genConsumeReg(op2);
+        emit->emitIns_R_R_FLAGS_COND(INS_ccmp, cmpSize, srcReg1, srcReg2, cflags, cond);
+    }
+
+    // Are we evaluating this into a register?
+    if (targetReg != REG_NA)
+    {
+        inst_SETCC(GenCondition::FromRelop(tree), tree->TypeGet(), targetReg);
+        genProduceReg(tree);
+    }
+}
 
-        // For the ccmp flags, get the condition of the compare.
-        insCflags cflags = InsCflagsForCcmp(InsCondForCompareOp(tree));
+//------------------------------------------------------------------------
+// genCodeForContainedCompareChain: Produce code for a chain of conditional compares.
+//
+// Only generates for contained nodes. Nodes that are not contained are assumed to be
+// generated as part of standard tree generation.
+//
+// Arguments:
+//    tree - the node. Either a compare or a tree of compares connected by ANDs.
+//    inchain - whether a contained chain is in progress.
+//    prev - If a chain is in progress, the condition of the previous compare.
+// Return:
+//    The last compare node generated.
+//
+void CodeGen::genCodeForContainedCompareChain(GenTreeOp* tree, bool* inchain, insCond* prevcond)
+{
+    if (tree->OperIs(GT_AND))
+    {
+        GenTreeOp* op1 = tree->gtGetOp1()->AsOp();
+        GenTreeOp* op2 = tree->gtGetOp2()->AsOp();
 
-        if (op2->isContainedIntOrIImmed())
+        if (!tree->isContained())
         {
-            GenTreeIntConCommon* intConst = op2->AsIntConCommon();
-            emit->emitIns_R_I_FLAGS_COND(INS_ccmp, cmpSize, srcReg1, (int)intConst->IconValue(), cflags, cond);
+            // An And that is not contained should not have any contained children.
+            assert(!op1->isContained() && !op2->isContained());
+            *inchain = false;
         }
         else
         {
-            regNumber srcReg2 = genConsumeReg(op2);
-            emit->emitIns_R_R_FLAGS_COND(INS_ccmp, cmpSize, srcReg1, srcReg2, cflags, cond);
+            // An And can only be contained if Op2 is contained.
+            assert(op2->isContained());
+
+            // If Op1 is contained, generate into flags. Otherwise, move the result into flags.
+            if (op1->isContained())
+            {
+                genCodeForContainedCompareChain(op1, inchain, prevcond);
+                assert(*inchain);
+            }
+            else
+            {
+                emitter* emit = GetEmitter();
+                emit->emitIns_R_I(INS_cmp, EA_ATTR(genTypeSize(op1)), op1->GetRegNum(), 1);
+                *prevcond = INS_COND_EQ;
```

Comment:
This should be changed to `!= 0`.

Reply:
Also don't we need to call `genConsumeReg` here?

Reply:
>Also don't we need to call genConsumeReg here?

I don't think so. op1 is not contained, so it's already been code generated into a register, and everything has already been consumed. We're then moving that result from the registers into flags so that op2 can use them.

Reply:
`genConsumeReg` needs to be called for uncontained operands. It is responsible for ensuring the value is available in the register that the operand was allocated into (e.g. by unspilling if necessary).